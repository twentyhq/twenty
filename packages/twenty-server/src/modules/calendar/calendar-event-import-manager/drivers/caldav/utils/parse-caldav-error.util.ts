import {
  CalendarEventImportDriverException,
  CalendarEventImportDriverExceptionCode,
} from 'src/modules/calendar/calendar-event-import-manager/drivers/exceptions/calendar-event-import-driver.exception';

export const parseCalDAVError = (
  error: Error,
): CalendarEventImportDriverException => {
  const { message } = error;

  switch (message) {
    case 'Collection does not exist on server':
      return new CalendarEventImportDriverException(
        message,
        CalendarEventImportDriverExceptionCode.NOT_FOUND,
      );

    case 'no account for smartCollectionSync':
    case 'no account for fetchAddressBooks':
    case 'no account for fetchCalendars':
    case 'Must have account before syncCalendars':
      return new CalendarEventImportDriverException(
        message,
        CalendarEventImportDriverExceptionCode.INSUFFICIENT_PERMISSIONS,
      );

    case 'cannot fetchVCards for undefined addressBook':
    case 'cannot find calendarUserAddresses':
    case 'cannot fetchCalendarObjects for undefined calendar':
    case 'cannot find homeUrl':
      return new CalendarEventImportDriverException(
        message,
        CalendarEventImportDriverExceptionCode.NOT_FOUND,
      );

    case 'Invalid credentials':
      return new CalendarEventImportDriverException(
        message,
        CalendarEventImportDriverExceptionCode.INSUFFICIENT_PERMISSIONS,
      );

    case 'Invalid auth method':
      return new CalendarEventImportDriverException(
        message,
        CalendarEventImportDriverExceptionCode.INSUFFICIENT_PERMISSIONS,
      );
  }

  return new CalendarEventImportDriverException(
    message,
    CalendarEventImportDriverExceptionCode.UNKNOWN,
  );
};
