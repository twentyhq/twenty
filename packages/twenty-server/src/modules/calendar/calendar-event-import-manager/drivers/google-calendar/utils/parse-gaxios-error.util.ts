import { GaxiosError } from 'gaxios';

import {
  CalendarEventImportDriverException,
  CalendarEventImportDriverExceptionCode,
} from 'src/modules/calendar/calendar-event-import-manager/drivers/exceptions/calendar-event-import-driver.exception';
import { MessageNetworkExceptionCode } from 'src/modules/messaging/message-import-manager/drivers/exceptions/message-network.exception';

export const parseGaxiosError = (
  error: GaxiosError,
): CalendarEventImportDriverException => {
  const { code } = error;

  switch (code) {
    case MessageNetworkExceptionCode.ECONNRESET:
    case MessageNetworkExceptionCode.ENOTFOUND:
    case MessageNetworkExceptionCode.ECONNABORTED:
    case MessageNetworkExceptionCode.ETIMEDOUT:
    case MessageNetworkExceptionCode.ERR_NETWORK:
      return new CalendarEventImportDriverException(
        error.message,
        CalendarEventImportDriverExceptionCode.TEMPORARY_ERROR,
      );

    default:
      return new CalendarEventImportDriverException(
        error.message,
        CalendarEventImportDriverExceptionCode.UNKNOWN_NETWORK_ERROR,
      );
  }
};
