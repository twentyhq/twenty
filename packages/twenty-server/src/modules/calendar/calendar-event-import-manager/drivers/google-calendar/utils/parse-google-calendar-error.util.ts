import {
  CalendarEventImportDriverException,
  CalendarEventImportDriverExceptionCode,
} from 'src/modules/calendar/calendar-event-import-manager/drivers/exceptions/calendar-event-import-driver.exception';

export const parseGoogleCalendarError = (error: {
  code?: number;
  reason: string;
  message: string;
}): CalendarEventImportDriverException => {
  const { code, reason, message } = error;

  switch (code) {
    case 400:
      if (reason === 'invalid_grant') {
        return new CalendarEventImportDriverException(
          message,
          CalendarEventImportDriverExceptionCode.INSUFFICIENT_PERMISSIONS,
        );
      }
      if (reason === 'failedPrecondition') {
        return new CalendarEventImportDriverException(
          message,
          CalendarEventImportDriverExceptionCode.TEMPORARY_ERROR,
        );
      }

      return new CalendarEventImportDriverException(
        message,
        CalendarEventImportDriverExceptionCode.UNKNOWN,
      );

    case 404:
      return new CalendarEventImportDriverException(
        message,
        CalendarEventImportDriverExceptionCode.NOT_FOUND,
      );

    case 429:
      return new CalendarEventImportDriverException(
        message,
        CalendarEventImportDriverExceptionCode.TEMPORARY_ERROR,
      );

    case 403:
      if (
        reason === 'rateLimitExceeded' ||
        reason === 'userRateLimitExceeded'
      ) {
        return new CalendarEventImportDriverException(
          message,
          CalendarEventImportDriverExceptionCode.TEMPORARY_ERROR,
        );
      } else {
        return new CalendarEventImportDriverException(
          message,
          CalendarEventImportDriverExceptionCode.INSUFFICIENT_PERMISSIONS,
        );
      }

    case 401:
      return new CalendarEventImportDriverException(
        message,
        CalendarEventImportDriverExceptionCode.INSUFFICIENT_PERMISSIONS,
      );
    case 500:
      if (reason === 'backendError' || reason === 'internal_failure') {
        return new CalendarEventImportDriverException(
          message,
          CalendarEventImportDriverExceptionCode.TEMPORARY_ERROR,
        );
      } else {
        return new CalendarEventImportDriverException(
          message,
          CalendarEventImportDriverExceptionCode.UNKNOWN,
        );
      }

    default:
      break;
  }

  return new CalendarEventImportDriverException(
    message,
    CalendarEventImportDriverExceptionCode.UNKNOWN,
  );
};
