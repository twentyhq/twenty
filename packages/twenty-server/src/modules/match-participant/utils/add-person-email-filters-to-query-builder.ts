import { SelectQueryBuilder } from 'typeorm';

import { PersonWorkspaceEntity } from 'src/modules/person/standard-objects/person.workspace-entity';

export interface AddPersonEmailFiltersToQueryBuilderOptions {
  queryBuilder: SelectQueryBuilder<PersonWorkspaceEntity>;
  emails: string[];
  excludePersonIds?: string[];
}

/**
 * Adds filters to a query builder to only return people with the given emails.
 * This is used to find people by their primary or additional emails.
 * We use the query builder here instead of the find method from typeorm
 * because we need to use the jsonb @> operator to check if the email is in the additional emails array.
 *
 * @param queryBuilder - The query builder to add the filters to
 * @param emails - The emails to filter by
 * @param excludePersonIds - The person IDs to exclude from the results
 */
export function addPersonEmailFiltersToQueryBuilder({
  queryBuilder,
  emails,
  excludePersonIds = [],
}: AddPersonEmailFiltersToQueryBuilderOptions): SelectQueryBuilder<PersonWorkspaceEntity> {
  const normalizedEmails = emails.map((email) => email.toLowerCase());

  queryBuilder = queryBuilder
    .select([
      'person.id',
      'person.emailsPrimaryEmail',
      'person.emailsAdditionalEmails',
    ])
    .where('LOWER(person.emailsPrimaryEmail) IN (:...emails)', {
      emails: normalizedEmails,
    });

  if (excludePersonIds.length > 0) {
    queryBuilder = queryBuilder.andWhere(
      'person.id NOT IN (:...excludePersonIds)',
      {
        excludePersonIds,
      },
    );
  }

  for (const [index, email] of normalizedEmails.entries()) {
    const emailParamName = `email${index}`;
    const orWhereIsInAdditionalEmail =
      excludePersonIds.length > 0
        ? `person.id NOT IN (:...excludePersonIds) AND person.emailsAdditionalEmails @> :${emailParamName}::jsonb`
        : `person.emailsAdditionalEmails @> :${emailParamName}::jsonb`;

    queryBuilder = queryBuilder.orWhere(orWhereIsInAdditionalEmail, {
      ...(excludePersonIds.length > 0 && { excludePersonIds }),
      [emailParamName]: JSON.stringify([email]),
    });
  }

  queryBuilder = queryBuilder.withDeleted();

  return queryBuilder;
}
