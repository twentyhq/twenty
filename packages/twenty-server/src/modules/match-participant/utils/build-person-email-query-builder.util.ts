import { SelectQueryBuilder } from 'typeorm';

import { PersonWorkspaceEntity } from 'src/modules/person/standard-objects/person.workspace-entity';

export interface BuildPersonEmailQueryBuilderOptions {
  queryBuilder: SelectQueryBuilder<PersonWorkspaceEntity>;
  emails: string[];
  excludePersonIds?: string[];
}

export function buildPersonEmailQueryBuilder({
  queryBuilder,
  emails,
  excludePersonIds = [],
}: BuildPersonEmailQueryBuilderOptions): SelectQueryBuilder<PersonWorkspaceEntity> {
  const normalizedEmails = emails.map((email) => email.toLowerCase());

  queryBuilder = queryBuilder
    .select([
      'person.id',
      'person.emailsPrimaryEmail',
      'person.emailsAdditionalEmails',
    ])
    .where('LOWER(person.emailsPrimaryEmail) IN (:...emails)', {
      emails: normalizedEmails,
    });

  if (excludePersonIds.length > 0) {
    queryBuilder = queryBuilder.andWhere(
      'person.id NOT IN (:...excludePersonIds)',
      {
        excludePersonIds,
      },
    );
  }

  for (const [index, email] of normalizedEmails.entries()) {
    const emailParamName = `email${index}`;
    const orWhereIsInAdditionalEmail =
      excludePersonIds.length > 0
        ? `person.id NOT IN (:...excludePersonIds) AND person.emailsAdditionalEmails @> :${emailParamName}::jsonb`
        : `person.emailsAdditionalEmails @> :${emailParamName}::jsonb`;

    queryBuilder = queryBuilder.orWhere(orWhereIsInAdditionalEmail, {
      ...(excludePersonIds.length > 0 && { excludePersonIds }),
      [emailParamName]: JSON.stringify([email]),
    });
  }

  queryBuilder = queryBuilder.withDeleted();

  return queryBuilder;
}
