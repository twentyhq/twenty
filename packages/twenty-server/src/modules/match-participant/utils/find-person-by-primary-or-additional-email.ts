import { PersonWorkspaceEntity } from 'src/modules/person/standard-objects/person.workspace-entity';

export const findPersonByPrimaryOrAdditionalEmail = ({
  people,
  email,
}: {
  people: PersonWorkspaceEntity[];
  email: string;
}): PersonWorkspaceEntity | undefined => {
  const lowercaseEmail = email.toLowerCase();

  const personWithPrimaryEmail = people.find(
    (person) => person.emails?.primaryEmail?.toLowerCase() === lowercaseEmail,
  );

  if (personWithPrimaryEmail) {
    return personWithPrimaryEmail;
  }

  const personWithAdditionalEmail = people.find((person) => {
    const additionalEmails = person.emails?.additionalEmails;

    if (!Array.isArray(additionalEmails)) {
      return false;
    }

    return additionalEmails.some(
      (additionalEmail) => additionalEmail.toLowerCase() === lowercaseEmail,
    );
  });

  return personWithAdditionalEmail;
};
