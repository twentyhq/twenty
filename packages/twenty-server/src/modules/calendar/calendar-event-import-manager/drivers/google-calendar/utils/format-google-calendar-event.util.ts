import { calendar_v3 as calendarV3 } from 'googleapis';

import { sanitizeCalendarEvent } from 'src/modules/calendar/calendar-event-import-manager/drivers/utils/sanitizeCalendarEvent';
import { CalendarEventParticipantResponseStatus } from 'src/modules/calendar/common/standard-objects/calendar-event-participant.workspace-entity';
import { FetchedCalendarEvent } from 'src/modules/calendar/common/types/fetched-calendar-event';

export const formatGoogleCalendarEvents = (
  events: calendarV3.Schema$Event[],
): FetchedCalendarEvent[] => {
  return events.map(formatGoogleCalendarEvent);
};

const formatGoogleCalendarEvent = (
  event: calendarV3.Schema$Event,
): FetchedCalendarEvent => {
  const formatResponseStatus = (status: string | null | undefined) => {
    switch (status) {
      case 'accepted':
        return CalendarEventParticipantResponseStatus.ACCEPTED;
      case 'declined':
        return CalendarEventParticipantResponseStatus.DECLINED;
      case 'tentative':
        return CalendarEventParticipantResponseStatus.TENTATIVE;
      default:
        return CalendarEventParticipantResponseStatus.NEEDS_ACTION;
    }
  };

  // Create the event object
  const calendarEvent: FetchedCalendarEvent = {
    title: event.summary ?? '',
    isCanceled: event.status === 'cancelled',
    isFullDay: event.start?.dateTime == null,
    startsAt: event.start?.dateTime ?? event.start?.date ?? '',
    endsAt: event.end?.dateTime ?? event.end?.date ?? '',
    id: event.id ?? '',
    externalCreatedAt: event.created ?? '',
    externalUpdatedAt: event.updated ?? '',
    description: event.description ?? '',
    location: event.location ?? '',
    iCalUID: event.iCalUID ?? '',
    conferenceSolution:
      event.conferenceData?.conferenceSolution?.key?.type ?? '',
    conferenceLinkLabel: event.conferenceData?.entryPoints?.[0]?.uri ?? '',
    conferenceLinkUrl: event.conferenceData?.entryPoints?.[0]?.uri ?? '',
    recurringEventExternalId: event.recurringEventId ?? '',
    participants:
      event.attendees?.map((attendee) => ({
        handle: attendee.email ?? '',
        displayName: attendee.displayName ?? '',
        isOrganizer: attendee.organizer === true,
        responseStatus: formatResponseStatus(attendee.responseStatus),
      })) ?? [],
    status: event.status ?? '',
  };

  const propertiesToSanitize: (keyof FetchedCalendarEvent)[] = [
    'title',
    'startsAt',
    'endsAt',
    'id',
    'externalCreatedAt',
    'externalUpdatedAt',
    'description',
    'location',
    'iCalUID',
    'conferenceSolution',
    'conferenceLinkLabel',
    'conferenceLinkUrl',
    'recurringEventExternalId',
    'status',
  ];

  return sanitizeCalendarEvent(calendarEvent, propertiesToSanitize);
};
