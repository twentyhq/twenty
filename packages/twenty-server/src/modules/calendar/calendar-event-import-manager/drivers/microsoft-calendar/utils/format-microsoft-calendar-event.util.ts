import {
  Event,
  NullableOption,
  ResponseType,
} from '@microsoft/microsoft-graph-types';

import { sanitizeCalendarEvent } from 'src/modules/calendar/calendar-event-import-manager/drivers/utils/sanitizeCalendarEvent';
import { CalendarEventParticipantResponseStatus } from 'src/modules/calendar/common/standard-objects/calendar-event-participant.workspace-entity';
import { CalendarEventWithParticipants } from 'src/modules/calendar/common/types/calendar-event';

export const formatMicrosoftCalendarEvents = (
  events: Event[],
): CalendarEventWithParticipants[] => {
  return events.map(formatMicrosoftCalendarEvent);
};

const formatMicrosoftCalendarEvent = (
  event: Event,
): CalendarEventWithParticipants => {
  const formatResponseStatus = (
    status: NullableOption<ResponseType> | undefined,
  ) => {
    switch (status) {
      case 'accepted':
      case 'organizer':
        return CalendarEventParticipantResponseStatus.ACCEPTED;
      case 'declined':
        return CalendarEventParticipantResponseStatus.DECLINED;
      case 'tentativelyAccepted':
        return CalendarEventParticipantResponseStatus.TENTATIVE;
      default:
        return CalendarEventParticipantResponseStatus.NEEDS_ACTION;
    }
  };

  const calendarEvent: CalendarEventWithParticipants = {
    title: event.subject ?? '',
    isCanceled: !!event.isCancelled,
    isFullDay: !!event.isAllDay,
    startsAt: event.start?.dateTime ?? null,
    endsAt: event.end?.dateTime ?? null,
    externalId: event.id ?? '',
    externalCreatedAt: event.createdDateTime ?? null,
    externalUpdatedAt: event.lastModifiedDateTime ?? null,
    description: event.body?.content ?? '',
    location: event.location?.displayName ?? '',
    iCalUID: event.iCalUId ?? '',
    conferenceSolution: event.onlineMeetingProvider ?? '',
    conferenceLinkLabel: event.onlineMeeting?.joinUrl ?? '',
    conferenceLinkUrl: event.onlineMeeting?.joinUrl ?? '',
    recurringEventExternalId: event.id ?? '',
    participants:
      event.attendees?.map((attendee) => ({
        handle: attendee.emailAddress?.address ?? '',
        displayName: attendee.emailAddress?.name ?? '',
        isOrganizer: attendee.status?.response === 'organizer',
        responseStatus: formatResponseStatus(attendee.status?.response),
      })) ?? [],
    status: '',
  };

  const propertiesToSanitize: (keyof CalendarEventWithParticipants)[] = [
    'title',
    'startsAt',
    'endsAt',
    'externalId',
    'externalCreatedAt',
    'externalUpdatedAt',
    'description',
    'location',
    'iCalUID',
    'conferenceSolution',
    'conferenceLinkLabel',
    'conferenceLinkUrl',
    'recurringEventExternalId',
    'status',
  ];

  return sanitizeCalendarEvent(calendarEvent, propertiesToSanitize);
};
