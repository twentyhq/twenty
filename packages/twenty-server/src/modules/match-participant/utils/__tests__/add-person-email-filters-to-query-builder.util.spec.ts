import { EachTestingContext } from 'twenty-shared/testing';
import { SelectQueryBuilder } from 'typeorm';

import { addPersonEmailFiltersToQueryBuilder } from 'src/modules/match-participant/utils/add-person-email-filters-to-query-builder';
import { PersonWorkspaceEntity } from 'src/modules/person/standard-objects/person.workspace-entity';

type AddPersonEmailFiltersToQueryBuilderTestCase = EachTestingContext<{
  emails: string[];
  excludePersonIds?: string[];
  description: string;
}>;

const testCases: AddPersonEmailFiltersToQueryBuilderTestCase[] = [
  {
    title: 'single email without exclusions',
    context: {
      emails: ['test@example.com'],
      description: 'should handle a single email address correctly',
    },
  },
  {
    title: 'multiple emails without exclusions',
    context: {
      emails: ['test@example.com', 'contact@company.com'],
      description: 'should handle multiple email addresses correctly',
    },
  },
  {
    title: 'single email with person ID exclusions',
    context: {
      emails: ['test@example.com'],
      excludePersonIds: ['person-1', 'person-2'],
      description: 'should handle exclusions with a single email',
    },
  },
  {
    title: 'multiple emails with exclusions',
    context: {
      emails: ['test@example.com', 'contact@company.com'],
      excludePersonIds: ['person-1'],
      description: 'should handle exclusions with multiple emails',
    },
  },
  {
    title: 'empty emails array',
    context: {
      emails: [],
      description: 'should handle empty email array gracefully',
    },
  },
  {
    title: 'emails with empty exclusion array',
    context: {
      emails: ['test@example.com'],
      excludePersonIds: [],
      description: 'should handle empty exclusions array',
    },
  },
  {
    title: 'three emails with unique parameter generation',
    context: {
      emails: [
        'email1@example.com',
        'email2@example.com',
        'email3@example.com',
      ],
      description: 'should create unique parameter names for each email',
    },
  },
  {
    title: 'case-insensitive email normalization',
    context: {
      emails: ['Test@Example.COM', 'CONTACT@Company.com'],
      description: 'should normalize emails to lowercase',
    },
  },
];

interface QueryBuilderCall {
  method: string;
  args: any[];
}

let queryBuilderCalls: QueryBuilderCall[] = [];

const mockQueryBuilder: Partial<SelectQueryBuilder<PersonWorkspaceEntity>> = {
  select: jest.fn().mockImplementation((...args) => {
    queryBuilderCalls.push({ method: 'select', args });

    return mockQueryBuilder;
  }),
  where: jest.fn().mockImplementation((...args) => {
    queryBuilderCalls.push({ method: 'where', args });

    return mockQueryBuilder;
  }),
  andWhere: jest.fn().mockImplementation((...args) => {
    queryBuilderCalls.push({ method: 'andWhere', args });

    return mockQueryBuilder;
  }),
  orWhere: jest.fn().mockImplementation((...args) => {
    queryBuilderCalls.push({ method: 'orWhere', args });

    return mockQueryBuilder;
  }),
  withDeleted: jest.fn().mockImplementation((...args) => {
    queryBuilderCalls.push({ method: 'withDeleted', args });

    return mockQueryBuilder;
  }),
};

describe('addPersonEmailFiltersToQueryBuilder', () => {
  beforeEach(() => {
    queryBuilderCalls = [];
    jest.clearAllMocks();
  });

  it.each(testCases)(
    '$title',
    ({ context: { emails, excludePersonIds, description } }) => {
      const result = addPersonEmailFiltersToQueryBuilder({
        queryBuilder:
          mockQueryBuilder as SelectQueryBuilder<PersonWorkspaceEntity>,
        emails,
        excludePersonIds,
      });

      expect(queryBuilderCalls).toMatchSnapshot(
        `${description} - query builder calls`,
      );

      expect(result).toBe(mockQueryBuilder);
    },
  );
});
