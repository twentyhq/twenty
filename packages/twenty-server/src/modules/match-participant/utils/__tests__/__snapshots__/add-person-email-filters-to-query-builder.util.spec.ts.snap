// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`addPersonEmailFiltersToQueryBuilder case-insensitive email normalization: should normalize emails to lowercase - query builder calls 1`] = `
[
  {
    "args": [
      [
        "person.id",
        "person.emailsPrimaryEmail",
        "person.emailsAdditionalEmails",
      ],
    ],
    "method": "select",
  },
  {
    "args": [
      "LOWER(person.emailsPrimaryEmail) IN (:...emails)",
      {
        "emails": [
          "test@example.com",
          "contact@company.com",
        ],
      },
    ],
    "method": "where",
  },
  {
    "args": [
      "person.emailsAdditionalEmails @> :email0::jsonb",
      {
        "email0": "["test@example.com"]",
      },
    ],
    "method": "orWhere",
  },
  {
    "args": [
      "person.emailsAdditionalEmails @> :email1::jsonb",
      {
        "email1": "["contact@company.com"]",
      },
    ],
    "method": "orWhere",
  },
  {
    "args": [],
    "method": "withDeleted",
  },
]
`;

exports[`addPersonEmailFiltersToQueryBuilder emails with empty exclusion array: should handle empty exclusions array - query builder calls 1`] = `
[
  {
    "args": [
      [
        "person.id",
        "person.emailsPrimaryEmail",
        "person.emailsAdditionalEmails",
      ],
    ],
    "method": "select",
  },
  {
    "args": [
      "LOWER(person.emailsPrimaryEmail) IN (:...emails)",
      {
        "emails": [
          "test@example.com",
        ],
      },
    ],
    "method": "where",
  },
  {
    "args": [
      "person.emailsAdditionalEmails @> :email0::jsonb",
      {
        "email0": "["test@example.com"]",
      },
    ],
    "method": "orWhere",
  },
  {
    "args": [],
    "method": "withDeleted",
  },
]
`;

exports[`addPersonEmailFiltersToQueryBuilder empty emails array: should handle empty email array gracefully - query builder calls 1`] = `
[
  {
    "args": [
      [
        "person.id",
        "person.emailsPrimaryEmail",
        "person.emailsAdditionalEmails",
      ],
    ],
    "method": "select",
  },
  {
    "args": [
      "LOWER(person.emailsPrimaryEmail) IN (:...emails)",
      {
        "emails": [],
      },
    ],
    "method": "where",
  },
  {
    "args": [],
    "method": "withDeleted",
  },
]
`;

exports[`addPersonEmailFiltersToQueryBuilder multiple emails with exclusions: should handle exclusions with multiple emails - query builder calls 1`] = `
[
  {
    "args": [
      [
        "person.id",
        "person.emailsPrimaryEmail",
        "person.emailsAdditionalEmails",
      ],
    ],
    "method": "select",
  },
  {
    "args": [
      "LOWER(person.emailsPrimaryEmail) IN (:...emails)",
      {
        "emails": [
          "test@example.com",
          "contact@company.com",
        ],
      },
    ],
    "method": "where",
  },
  {
    "args": [
      "person.id NOT IN (:...excludePersonIds)",
      {
        "excludePersonIds": [
          "person-1",
        ],
      },
    ],
    "method": "andWhere",
  },
  {
    "args": [
      "person.id NOT IN (:...excludePersonIds) AND person.emailsAdditionalEmails @> :email0::jsonb",
      {
        "email0": "["test@example.com"]",
        "excludePersonIds": [
          "person-1",
        ],
      },
    ],
    "method": "orWhere",
  },
  {
    "args": [
      "person.id NOT IN (:...excludePersonIds) AND person.emailsAdditionalEmails @> :email1::jsonb",
      {
        "email1": "["contact@company.com"]",
        "excludePersonIds": [
          "person-1",
        ],
      },
    ],
    "method": "orWhere",
  },
  {
    "args": [],
    "method": "withDeleted",
  },
]
`;

exports[`addPersonEmailFiltersToQueryBuilder multiple emails without exclusions: should handle multiple email addresses correctly - query builder calls 1`] = `
[
  {
    "args": [
      [
        "person.id",
        "person.emailsPrimaryEmail",
        "person.emailsAdditionalEmails",
      ],
    ],
    "method": "select",
  },
  {
    "args": [
      "LOWER(person.emailsPrimaryEmail) IN (:...emails)",
      {
        "emails": [
          "test@example.com",
          "contact@company.com",
        ],
      },
    ],
    "method": "where",
  },
  {
    "args": [
      "person.emailsAdditionalEmails @> :email0::jsonb",
      {
        "email0": "["test@example.com"]",
      },
    ],
    "method": "orWhere",
  },
  {
    "args": [
      "person.emailsAdditionalEmails @> :email1::jsonb",
      {
        "email1": "["contact@company.com"]",
      },
    ],
    "method": "orWhere",
  },
  {
    "args": [],
    "method": "withDeleted",
  },
]
`;

exports[`addPersonEmailFiltersToQueryBuilder single email with person ID exclusions: should handle exclusions with a single email - query builder calls 1`] = `
[
  {
    "args": [
      [
        "person.id",
        "person.emailsPrimaryEmail",
        "person.emailsAdditionalEmails",
      ],
    ],
    "method": "select",
  },
  {
    "args": [
      "LOWER(person.emailsPrimaryEmail) IN (:...emails)",
      {
        "emails": [
          "test@example.com",
        ],
      },
    ],
    "method": "where",
  },
  {
    "args": [
      "person.id NOT IN (:...excludePersonIds)",
      {
        "excludePersonIds": [
          "person-1",
          "person-2",
        ],
      },
    ],
    "method": "andWhere",
  },
  {
    "args": [
      "person.id NOT IN (:...excludePersonIds) AND person.emailsAdditionalEmails @> :email0::jsonb",
      {
        "email0": "["test@example.com"]",
        "excludePersonIds": [
          "person-1",
          "person-2",
        ],
      },
    ],
    "method": "orWhere",
  },
  {
    "args": [],
    "method": "withDeleted",
  },
]
`;

exports[`addPersonEmailFiltersToQueryBuilder single email without exclusions: should handle a single email address correctly - query builder calls 1`] = `
[
  {
    "args": [
      [
        "person.id",
        "person.emailsPrimaryEmail",
        "person.emailsAdditionalEmails",
      ],
    ],
    "method": "select",
  },
  {
    "args": [
      "LOWER(person.emailsPrimaryEmail) IN (:...emails)",
      {
        "emails": [
          "test@example.com",
        ],
      },
    ],
    "method": "where",
  },
  {
    "args": [
      "person.emailsAdditionalEmails @> :email0::jsonb",
      {
        "email0": "["test@example.com"]",
      },
    ],
    "method": "orWhere",
  },
  {
    "args": [],
    "method": "withDeleted",
  },
]
`;

exports[`addPersonEmailFiltersToQueryBuilder three emails with unique parameter generation: should create unique parameter names for each email - query builder calls 1`] = `
[
  {
    "args": [
      [
        "person.id",
        "person.emailsPrimaryEmail",
        "person.emailsAdditionalEmails",
      ],
    ],
    "method": "select",
  },
  {
    "args": [
      "LOWER(person.emailsPrimaryEmail) IN (:...emails)",
      {
        "emails": [
          "email1@example.com",
          "email2@example.com",
          "email3@example.com",
        ],
      },
    ],
    "method": "where",
  },
  {
    "args": [
      "person.emailsAdditionalEmails @> :email0::jsonb",
      {
        "email0": "["email1@example.com"]",
      },
    ],
    "method": "orWhere",
  },
  {
    "args": [
      "person.emailsAdditionalEmails @> :email1::jsonb",
      {
        "email1": "["email2@example.com"]",
      },
    ],
    "method": "orWhere",
  },
  {
    "args": [
      "person.emailsAdditionalEmails @> :email2::jsonb",
      {
        "email2": "["email3@example.com"]",
      },
    ],
    "method": "orWhere",
  },
  {
    "args": [],
    "method": "withDeleted",
  },
]
`;
